### ISSUE TITLE
limit simultaneous downloads per hoster & switch to different hoster if one is "blocked"

### ISSUE THREAD

**Issue by igel-kun** [pyload/pyload#255]
I stick to free downloads usually, but most hosters' free service doesn't allow more than 1 simultaneous download. What I'd love is to set a "limit" such that from each hoster, at most "limit" downloads are active at all times.

Furthermore (this may be a different issue and I should open another ticket, not sure, what do you think?), when some download requires wait time, pyload could maybe download a file from a different hoster in the meantime.
So what I have in mind is this queue:

myqueue.fill:
while( self is not filled && global queue is not empty){
url = top of global queue
hoster = get hoster from url
if ( used slots[hoster] < limit[hoster] ){
push url to self
slots[hoster]++
}
}

If some url turns out to have to wait more than some threshold of time, then it waives its spot and a new spot opens at the end of the queue:

myqueue.waive spot (spot_to_waive):
push url of spot_to_waive to global queue
delete spot_to_waive from this
hoster = get hoster from url
slots[hoster]--
myqueue.fill

An option may be necessary to let the user exclude some hosters from this process (mainly those that require a captcha in order to find out the wait time)

**Comment by igel-kun** [comment]
this is still valid, but I can maybe discribe it a little better:
I'd like an option to mark hosters "busy" for a certain time such that no URL belonging to a busy hoster is treated. Let's have an example: Imagine you have 3 URLs in the queue, 2 from hoster a, 1 from hoster b. Your simultanous download limit is 1.

## current behavior:
1st URL loads, then hoster a goes into "wait for 2h before you can download again"-mode so the second URL from hoster a blocks for 2h in "waiting" state.

## expected behavior:
1st URL loads, then hoster a goes into "wait for 2h before you can download again"-mode _which_is_detected_by_pyload_ and a is set to "busy" for 2h. pyload starts up the download of hoster b in the meantime.

**Comment by igel-kun** [comment]
it could be some kind of config option x telling pyload to set a hoster "busy" if a wait for more than x is requested

**Comment by GammaC0de** [comment]
What you are asking is to change the logic of pyLoad's scheduler which is something that have to be done with extreme cousion..

**Comment by GammaC0de** [comment]
Can you give me a hoster that you can download multiple files and have this time limitation?

**Comment by igel-kun** [comment]
I'm not sure I understand. You can download multiple files from any hoster. Just not at the same time, but "at the same time" is also not necessary for the scenario I'm describing above. So, I guess our favorite Keep2Share is an example for such a hoster.

**Comment by GammaC0de** [comment]
I have a working prototype, would you like to test?

**Comment by igel-kun** [comment]
yes, I'm interested in testing that - could take a little while 'cause my connection is broken right now...

**Comment by GammaC0de** [comment]
Try this patch:
[255.patch.zip](https://github.com/pyload/pyload/files/3294432/255.patch.zip)

**Comment by nitedreemer** [comment]
Hi GammaC0de

I have just discovered a similar issue in python with Nitroflare hoster. 

In free mode this hoster allows you to download 1 file per 180 min. After this "180 min blocking period" you can start downloading another file - disregard whether the previous file has already been fully downloaded or not.

But pyload does not start a new download for me after this "180 min blocking period" - neither from existing queue nor if I add a new one manually.

Does the patch address this as well ?

**Comment by GammaC0de** [comment]
@nitedreemer Please open a new issues, thanks.
